Introduce temp watch protection for the bed

scalez opened the pull request:
When setting the bed temp via `M140`/`M190`, if the thermistor does not read an increase of `WATCH_BED_TEMP_INCREASE` degrees within `WATCH_BED_TEMP_PERIOD` seconds then it will throw "`Error:Heating failed, system stopped! Heater_ID: bed`" and call the `kill()` function.

scalez committed:
Introduce temp watch protection for the bed

When setting the bed temp via M140/M190 if the thermistor does not read an increase of WATCH_BED_TEMP_INCREASE degrees by WATCH_BED_TEMP_PERIOD seconds then it will throw "Error:Heating failed, system stopped! Heater_ID: bed" and call the kill() function.

Conflicts:
	Marlin/Configuration_adv.h

scalez committed:
Use WATCH_BED_TEMP_PERIOD and WATCH_BED_TEMP_INCREASE in the example configuration headers and not WATCH_TEMP_PERIOD and WATCH_TEMP_INCREASE

AnHardt commented:
This looks like a well made reinvention of my #2310 and it's followers.
If we don't have a maximum test temperature here, we will get a lot of false positives 
(https://github.com/MarlinFirmware/Marlin/pull/2310#issuecomment-116054413)

To find a consensus about the activation of this feature will be much easier these days.

Most of the beds out there do not need this feature at all, because they are self limiting and there is no risk of fire. But some (me) desperately need it for their monster-bed-heaters:
![picture 162](https://cloud.githubusercontent.com/assets/211931/14616036/ea1a7b80-05a8-11e6-87af-846bc6a417a8.jpg)
![picture 163](https://cloud.githubusercontent.com/assets/211931/14616041/f0cff13a-05a8-11e6-8454-2235f07cec9c.jpg)

If merged in the RC-phase i recommend a deactivated by default version.

scalez committed:
Properly type cast watch_bed_next_ms's assignment

scalez commented:
@AnHardt Not sure how I missed the fact that `millis_t` is typedef'ed as an `unsigned long` but now `WATCH_BED_TEMP_PERIOD` should be able to hold a value up to 4294967 seconds (49.71 days)! I set  `WATCH_BED_TEMP_PERIOD` to 60 seconds, I'd say that there should be a point where we can claim that if a bed does not reach its target temp within an extreme `WATCH_BED_TEMP_PERIOD` then the design is inherently flawed and the designer/user should be aware that what they're doing is potentially dangerous.

jbrazio commented:
@AnHardt will you validate this PR with your ~1KW setup ? :-)

AnHardt commented:
@jbrazio 
Sorry. No. No in the near future (days - weeks). Since Easter I'm still barely healthy enough to keep my business going and to drop a few lines now and then.

But as long the algorithm is the same as in the nozzle heater version, i have unlimited trust in it.
@scalez 
Alone finding reasonable default values for the case the target temperature is above the self-limiting-temperature, is an impossible task. If the bed will not reach the target temperature the user will realize that - it does not need a error message and a safety shutdown then and more important: "There is no risk of fire" . For that a maximum test temperature is needed and that has to be well below the self-limiting-temperature - if there is a self-limiting-temperature. If the is no self-limiting before self destruction -  the maximum test temperature can be MAX_TEMP_BED.).

scalez commented:
@AnHardt sincerely hope you get better.

The logic is identical to the nozzle.
If the designer/user is aware of the self-limiting-temperature then the case when the target is unreachable should never happen. If the designer/user is setting the target temp using M140 and printing prior to the target temperature being reach (i.e. using M190) then that can also be considered a design/user error.

AnHardt commented:
In the days when i begun to experiment with, heat activated epoxy's, on high temperature beds, before i got the power-beds, i used to say "give me all you have", give me 150°, even if the bed barely could reach 225° maybe 130° with closed printer doors.
If i remember correct, for an original K8200, "give me all" is the standard for printing ABS.

Of course this is not what you want, but the best you can get, and more common than you might think.

thinkyhead commented:
This feature and the Bed Temp Residency checking have both been hovering on the periphery as features "we never wanted to add" because heated beds operate in a lower temp range and are –as pointed out– pretty well self-limiting. Still, for symmetry I don't think it's bad to have these two features added now. But I'm sure we can leave them disabled by default.

We ought to be leaving these extra bed temperature features (and a couple of other enhancements disabled by default) anyway, to accommodate the smaller (64K) boards that are now getting crowded out by the ballooning size of the Marlin binary.

With that in mind, keep an eye out for any chance to reduce the size of the "standard" or "default" install of Marlin. Aim for a target binary as small as version 1.0.2-1, or even smaller if possible.

jbrazio commented:
There is no way to only disable this feature, it is enabled by `THERMAL_PROTECTION_BED` which currently controls `THERMAL_PROTECTION_BED_HYSTERESIS`.. nothing against, just a reminder.

Blue-Marlin commented:
There are several ways to check that separately.
On of them is to check for

```
#if WATCH_BED_TEMP_PERIOD > 0
```

instead of

```
#if ENABLED(THERMAL_PROTECTION_BED)
```

scalez committed:
Merge remote-tracking branch 'upstream/RCBugFix' into BedTempWatch

Conflicts:
	Marlin/example_configurations/delta/kossel_xl/Configuration_adv.h

scalez commented:
@thinkyhead @Blue-Marlin I can wrap the heating sanity check code for both the nozzle and the bed with `#if WATCH_TEMP_PERIOD > 0` and `#if WATCH_BED_TEMP_PERIOD > 0`, and set `#define WATCH_BED_TEMP_PERIOD 0` in the PR if we get a consensus. Let me know what should be done and I can commit it. Ultimately, I believe a new macro should be created for `> 0` purposes.

jbrazio commented:
NVM that will [in the long run] be worst. Let's keep the config flags for the bed as close as possible as the ones for the nozzle.

As this is a safety feature, do we have any idea what would be a good fits-the-most default value ? Let's do some testing if nothing wrong comes then let's ship this enabled by default as we do for nozzle.

thinkyhead commented:
For "completeness" I suppose it makes sense to wrap the appropriate code in conditionals like the following, so we can disable these "sub-features" by either non-defining or setting the option to zero:

``` cpp
#if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```

We already have some precedent for this, as it allows us to avoid needing to add another "enable" flag.

thinkyhead commented on Marlin/example_configurations/tvrrug/Round2/Configuration_adv.h:
```diff
@@ -87,6 +87,18 @@
+   * WATCH_TEMP_INCREASE should not be below 2.
+   */
+  #define WATCH_TEMP_PERIOD 20                // Seconds
+  #define WATCH_TEMP_INCREASE 2               // Degrees Celsius
```
These lines should be in the block above, and the `WATCH_BED_TEMP_*` items in the code above should be here.

scalez committed:
Use WATCH_TEMP_PERIOD > 0 and WATCH_BED_TEMP_PERIOD > 0 in each preprocessor conditional involving a check for whether or not they are enabled.

start_watching_bed() is now also called in ultralcd.cpp

Also correct SanityCheck error for THERMAL_PROTECTION_BED.

scalez committed:
Correct the locations of watch temp definitions for the hotend and bed in tvrrug Round2's example Configuration_adv.h

scalez commented:
@thinkyhead @jbrazio check out my last two commits, I think it should be good to go now.

jbrazio commented:
@scalez thanks, it's looking good !

scalez committed:
It's okay for WATCH_TEMP_PERIOD to be above 32 since watch_heater_next_ms is now defined as millis_t and millis_t is typedef'ed as an unsigned long not int (WATCH_TEMP_PERIOD's limit is therefore 4294967s).

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -439,9 +439,9 @@
 #if WATCH_TEMP_PERIOD > 500
   #error WATCH_TEMP_PERIOD now uses seconds instead of milliseconds.
-#elif DISABLED(THERMAL_PROTECTION_HOTENDS) && (defined(WATCH_TEMP_PERIOD) || defined(THERMAL_PROTECTION_PERIOD))
+#elif (DISABLED(THERMAL_PROTECTION_HOTENDS) || WATCH_TEMP_PERIOD <= 0) && (defined(WATCH_TEMP_PERIOD) || (defined(THERMAL_PROTECTION_PERIOD) && WATCH_TEMP_PERIOD > 0))
```
The original reason for these tests was to deal with older configurations that didn't have `THERMAL_PROTECTION_HOTENDS` or `THERMAL_PROTECTION_BED` options, so the other options could be enabled without them. This isn't a problem with your new options, which are … new. So you can drop all your changes to `SanityCheck.h` here.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -97,14 +97,14 @@ unsigned char soft_pwm_bed;
 #endif
 
-#if ENABLED(THERMAL_PROTECTION_HOTENDS) || ENABLED(THERMAL_PROTECTION_BED)
+#if (ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0) || (ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0)
```
The `thermal_runaway_protection` function isn't related to `WATCH_TEMP_PERIOD` so you shouldn't be checking it here.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -97,14 +97,14 @@ unsigned char soft_pwm_bed;
   enum TRState { TRReset, TRInactive, TRFirstHeating, TRStable, TRRunaway };
   void thermal_runaway_protection(TRState* state, millis_t* timer, float temperature, float target_temperature, int heater_id, int period_seconds, int hysteresis_degc);
-  #if ENABLED(THERMAL_PROTECTION_HOTENDS)
+  #if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
Same here and below. This condition will break compilation if `WATCH_TEMP_PERIOD` is zero.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -200,11 +200,16 @@ static float analog2temp(int raw, uint8_t e);
 static void updateTemperaturesFromRawValues();
 
-#if ENABLED(THERMAL_PROTECTION_HOTENDS)
+#if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
This is ok however.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -657,14 +662,14 @@ void manage_heater() {
   for (int e = 0; e < EXTRUDERS; e++) {
 
-    #if ENABLED(THERMAL_PROTECTION_HOTENDS)
+    #if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
Not here.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -728,7 +751,7 @@ void manage_heater() {
   #if TEMP_SENSOR_BED != 0
 
-    #if ENABLED(THERMAL_PROTECTION_BED)
+    #if ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0
```
No `WATCH_*_TEMP_PERIOD` check here or below.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1116,7 +1139,23 @@ void tp_init() {
+  }
+#endif
+
+#if (ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0) || (ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0)
```
This line should be unchanged too.

thinkyhead commented:
Other than the needed changes I just listed, it all looks good!

scalez committed:
Only use WATCH_*TEMP_PERIOD where needed

scalez commented:
@thinkyhead https://github.com/MarlinFirmware/Marlin/pull/3554/commits/abc3bc18afd634fe311801cfbac1fef2ed101dd6 made the needed changes, let me know what you think.

thinkyhead commented:
@scalez I've squashed all the commits into a single commit and will merge #3653 shortly.
